### DISCUSSION TITLE
Exo Composition: how to enhance an exo object with additional capabilities / methods?

### DISCUSSION THREAD

**Discussion by 0xpatrickdev** [Agoric/agoric-sdk#9035] [Q&A] [answered]
There is a JS convention where a `with*()` or `enhance*()` function takes an object and adds additional functionality to it:

```js
const burger = { bread: true, tomato: true, patty: true };
const withCheese = foo => ({ ...foo, cheese: true });
const cheeseburger = withCheese(burger);
```

If I want to do the same with `zone.exoClass()` or `zone.exoClassKit()`, is this possible? Any gotchas I should look out for?

Edit, keywords for search: mixin, higher order function, compose, composition

**Comment by dckc** [comment, marked as answer]
_briefly_: use [prepareGuardedAttenuator](https://github.com/Agoric/agoric-sdk/blob/master/packages/internal/src/callback.js#L315), which includes:

```
   * Create an exo object whose behavior is composed from a default target
   * and/or individual method override callbacks.
```

By way of example, `myAddressNameAdmin` adds a `provideChild` method to enhance the normal `nameAdmin` from `prepareNameHubKit`.

https://github.com/Agoric/agoric-sdk/blob/c5284e4721ad2702c4551f216bb8eed7abe4d7cf/packages/vats/src/vat-provisioning.js#L20
